File: branches/Lite-1_0/src/main/org/xidea/el/parser/ExpressionTokenizer.java
package org.xidea.el.parser;

import static org.xidea.el.ExpressionToken.OP_ADD;
import static org.xidea.el.ExpressionToken.OP_AND;
import static org.xidea.el.ExpressionToken.OP_GET_PROP;
import static org.xidea.el.ExpressionToken.OP_INVOKE_METHOD;
import static org.xidea.el.ExpressionToken.OP_MAP_PUSH;
import static org.xidea.el.ExpressionToken.OP_NEG;
import static org.xidea.el.ExpressionToken.OP_OR;
import static org.xidea.el.ExpressionToken.OP_PARAM_JOIN;
import static org.xidea.el.ExpressionToken.OP_POS;
import static org.xidea.el.ExpressionToken.OP_QUESTION;
import static org.xidea.el.ExpressionToken.OP_QUESTION_SELECT;
import static org.xidea.el.ExpressionToken.OP_STATIC_GET_PROP;
import static org.xidea.el.ExpressionToken.OP_SUB;
import static org.xidea.el.ExpressionToken.VALUE_CONSTANTS;
import static org.xidea.el.ExpressionToken.VALUE_LAZY;
import static org.xidea.el.ExpressionToken.VALUE_NEW_LIST;
import static org.xidea.el.ExpressionToken.VALUE_NEW_MAP;
import static org.xidea.el.ExpressionToken.VALUE_VAR;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

import org.xidea.el.ExpressionSyntaxException;
import org.xidea.el.ExpressionToken;
import org.xidea.el.fn.NumberArithmetic;
import org.xidea.el.json.JSONTokenizer;

public class ExpressionTokenizer extends JSONTokenizer {

	//编译期间标记，compile time object
	static final int BRACKET_BEGIN = 0xFFFE;//([{;
	static final int BRACKET_END = 0xFFFF;//)]};
	
	private static enum Status {
		STATUS_BEGIN, STATUS_EXPRESSION, STATUS_OPERATOR
	}
	private static NumberArithmetic na = new NumberArithmetic();

	private Status status = Status.STATUS_BEGIN;
	private int previousType = Integer.MIN_VALUE;

	protected ArrayList<ExpressionToken> tokens = new ArrayList<ExpressionToken>();
	protected List<ExpressionToken> expressions;

	public ExpressionTokenizer(String value) {
		super(value);
		parseEL();
		this.expressions = right(this.tokens.iterator());
	}

	private int getPriority(int type) {
		switch (type) {
		case BRACKET_BEGIN:
		case BRACKET_END:
			return Integer.MIN_VALUE;
		default:
			return type & 30;
		}
	}

	private boolean rightEnd(ExpressionToken item, ExpressionToken privious) {
		int t1 = privious.getType();
		int t2 = item.getType();
		int p1 = getPriority(t1);
		int p2 = getPriority(t2);
		//1?1:3 + 0?5:7 ==>1
		//1?0?5:7:3 ==>7
		//1?0?5:0?11:13:3 ==>13
		if(p2 <= p1){
			if(p2 == p1){
				if(t2 == OP_QUESTION_SELECT){
					return t1 == OP_QUESTION;
				}else if(t2 == OP_QUESTION){
					return t1 != OP_QUESTION_SELECT;
				}
			}
			return true;
		}else{
			return false;
		}
	}

	// 将中序表达式转换为右序表达式
	private List<ExpressionToken> right(Iterator<ExpressionToken> tokens) {
		LinkedList<List<ExpressionToken>> rightStack = new LinkedList<List<ExpressionToken>>();
		rightStack.addFirst(new ArrayList<ExpressionToken>()); // 存储右序表达式

		LinkedList<ExpressionToken> buffer = new LinkedList<ExpressionToken>();

		while (tokens.hasNext()) {
			final ExpressionToken item = tokens.next();
			if (item.getType() > 0) {
				if (buffer.isEmpty()) {
					buffer.addFirst(item);
				} else if (item.getType() == BRACKET_BEGIN) {// ("(")
					buffer.addFirst(item);
				} else if (item.getType() == BRACKET_END) {// .equals(")"))
					while (true) {
						ExpressionToken operator = buffer.removeFirst();
						if (operator.getType() == BRACKET_BEGIN) {
							break;
						}
						addRightOperator(rightStack, operator);
					}
				} else {
					while (!buffer.isEmpty()
							&& rightEnd(item, buffer.getFirst())) {
						ExpressionToken operator = buffer.removeFirst();
						// if (operator.getType() !=
						// BRACKET_BEGIN){
						addRightOperator(rightStack, operator);
					}
					buffer.addFirst(item);
				}
			} else {// lazy begin value exp
				addRightToken(rightStack, item);
			}
		}
		while (!buffer.isEmpty()) {
			ExpressionToken operator = buffer.removeFirst();
			addRightOperator(rightStack, operator);
		}
		return rightStack.getFirst();
	}

	private void addRightOperator(LinkedList<List<ExpressionToken>> rightStack,
			ExpressionToken operator) {
		switch (operator.getType()) {
		case OP_OR:
		case OP_AND:
		case OP_QUESTION:
		case OP_QUESTION_SELECT:
			List<ExpressionToken> children = rightStack.removeFirst();
			List<ExpressionToken> list = rightStack.getFirst();
			if (children.size() == 1) {
				list.set(list.size() - 1, children.get(0));
				break;
			}
			if (children.size() == 2) {
				ExpressionToken t1 = children.get(0);
				ExpressionToken t2 = children.get(1);
				if (t2.getType() == VALUE_CONSTANTS
						&& t2.getParam() instanceof Number) {
					if (t1.getType() == OP_POS) {
						list.set(list.size() - 1, t2);
						break;
					} else if (t1.getType() == OP_NEG) {
						t2 = new TokenImpl(OP_NEG, na.subtract(0,(Number)t2.getParam()));
						list.set(list.size() - 1, t2);
						break;
					}
				}
			}
			ExpressionToken token = (ExpressionToken) list.get(list.size() - 1);
			((TokenImpl) token).setParam(reverseArray(children));

		}
		addRightToken(rightStack, operator);
	}

	private void addRightToken(LinkedList<List<ExpressionToken>> rightStack,
			ExpressionToken token) {
		List<ExpressionToken> list = rightStack.getFirst();
		if (token.getType() == OP_GET_PROP) {
			int last = list.size() - 1;
			if (last >= 0) {
				ExpressionToken previous = list.get(last);
				if (previous.getType() == VALUE_CONSTANTS) {
					list.remove(last);
					token = new TokenImpl(OP_STATIC_GET_PROP, previous
							.getParam());
				}
			}
		} else if (token.getType() == VALUE_LAZY) {
			rightStack.addFirst(new ArrayList<ExpressionToken>());
		}
		list.add(token);
	}

	private ExpressionToken[] reverseArray(List<ExpressionToken> list) {
		ExpressionToken[] expression = new ExpressionToken[list.size()];
		int i = expression.length - 1;
		for (ExpressionToken expressionToken : list) {
			expression[i--] = expressionToken;
		}
		return expression;
	}

	public Tokens getTokens() {
		return new Tokens(reverseArray(expressions));
	}

	protected void parseEL() {
		skipSpace(0);
		while (start < end) {
			char c = value.charAt(start);
			if (c == '"' || c == '\'') {
				String text = findString();
				addKeyOrObject(text, false);
			} else if (c >= '0' && c <= '9') {
				Number number = findNumber();
				addKeyOrObject(number, false);
			} else if (Character.isJavaIdentifierStart(c)) {
				String id = findId();
				ExpressionToken constains = TokenImpl.getConstainsToken(id);
				if (constains == null) {
					skipSpace(0);
					if (previousType == OP_GET_PROP) {
						addToken(new TokenImpl(VALUE_CONSTANTS, id));
					} else {
						addKeyOrObject(id, true);
					}
				} else {
					addToken(constains);
				}
			} else {
				String op = findOperator();
				// if (value.startsWith(op, start))
				if (op != null) {
					parseOperator(op);
				} else {
					throw new ExpressionSyntaxException("语法错误:" + value + "@"
							+ start);
				}
			}
			skipSpace(0);
		}
	}

	private String findOperator() {// optimize json ,:[{}]
		switch (value.charAt(start)) {
		case '!':// !,!=
		case '>':// >,>=
		case '<':// <,<=
			if (value.charAt(start + 1) == '=') {
				return value.substring(start, start += 2);
			}
		case ',':// optimize for json
		case ':':// 3op,map key
		case '[':// list
		case ']':
		case '{':// map
		case '}':
		case '(':// quote
		case ')':
		case '.':// prop
		case '?':// 3op
		case '+':// 5op
		case '-':
		case '*':
		case '/':
		case '%':
			return value.substring(start, start += 1);

		case '=':// ==
		case '&':// &&
		case '|':// ||
			assert (value.charAt(start) == value.charAt(start + 1));
			return value.substring(start, start += 2);
		}
		return null;
	}

	/**
	 * 碰見:和,的時候，就需要檢查是否事map的間隔符號了
	 * 
	 * @return
	 */
	private boolean isMapMethod() {
		int i = tokens.size() - 1;
		int depth = 0;
		for (; i >= 0; i--) {
			ExpressionToken token = tokens.get(i);
			int type = token.getType();
			if (depth == 0) {
				if (type == OP_MAP_PUSH || type == VALUE_NEW_MAP) {// (
					// <#newMap>
					// <#push>
					return true;
				} else if (type == OP_PARAM_JOIN) {// (
					// <#newList>
					// <#param_join>
					return false;
				}
			}
			if (type == BRACKET_BEGIN) {
				depth--;
			} else if (type == BRACKET_END) {
				depth++;
			}
		}
		return false;
	}

	private void parseOperator(String op) {
		if (op.length() == 1) {
			switch (op.charAt(0)) {
			case '(':
				if (status == Status.STATUS_EXPRESSION) {
					addToken(new TokenImpl(OP_INVOKE_METHOD, null));
					if (skipSpace(')')) {
						addToken(new TokenImpl(VALUE_CONSTANTS,
								Collections.EMPTY_LIST));
						start++;
					} else {
						addList();
					}

				} else {
					addToken(new TokenImpl(BRACKET_BEGIN, null));
				}
				break;
			case '[':
				if (status == Status.STATUS_EXPRESSION) {// getProperty
					addToken(new TokenImpl(OP_GET_PROP, null));
					addToken(new TokenImpl(BRACKET_BEGIN, null));
				} else {// list
					addList();
				}
				break;
			case '{':
				addMap();
				break;
			case '}':
			case ']':
			case ')':
				addToken(new TokenImpl(BRACKET_END, null));
				break;
			case '+'://
				addToken(new TokenImpl(
						status == Status.STATUS_EXPRESSION ? OP_ADD : OP_POS,
						null));
				// addToken(OperatorToken.getToken(SKIP_AND));
				break;
			case '-':
				addToken(new TokenImpl(
						status == Status.STATUS_EXPRESSION ? OP_SUB : OP_NEG,
						null));
				// addToken(OperatorToken.getToken(SKIP_AND));
				break;
			case '?':// ?:
				addToken(new TokenImpl(OP_QUESTION, null));
				// addToken(OperatorToken.getToken(SKIP_QUESTION));
				addToken(new TokenImpl(VALUE_LAZY, null));
				break;
			case ':':// :(object_setter is skiped)
				addToken(new TokenImpl(OP_QUESTION_SELECT, null));
				addToken(new TokenImpl(VALUE_LAZY, null));
				break;
			case ',':// :(object_setter is skiped,',' should
				// be skip)
				if (!isMapMethod()) {
					addToken(new TokenImpl(OP_PARAM_JOIN, null));

				}
				break;
			case '/':
				char next = value.charAt(start);
				if (next == '/' || next == '*') {
					start--;
					skipComment();
					break;
				}
			default:
				addToken(new TokenImpl(TokenImpl.findType(op), null));
			}
		} else if (op.equals("||")) { // ||
			addToken(new TokenImpl(OP_OR, null));
			addToken(new TokenImpl(VALUE_LAZY, null));
			// addToken(LazyToken.LAZY_TOKEN_END);
		} else if (op.equals("&&")) {// &&
			addToken(new TokenImpl(OP_AND, null));
			addToken(new TokenImpl(VALUE_LAZY, null));
			// addToken(OperatorToken.getToken(SKIP_AND));
		} else {
			addToken(new TokenImpl(TokenImpl.findType(op), null));
		}

	}

	private void addToken(ExpressionToken token) {
		switch (token.getType()) {
		case BRACKET_BEGIN:
			status = Status.STATUS_BEGIN;
			break;
		case VALUE_CONSTANTS:
		case VALUE_VAR:
		case BRACKET_END:
			status = Status.STATUS_EXPRESSION;
			break;
		default:
			status = Status.STATUS_OPERATOR;
			break;
		}
		// previousType2 = previousType;
		previousType = token.getType();
		tokens.add(token);
	}

	private void addKeyOrObject(Object object, boolean isVar) {
		if (skipSpace(':') && isMapMethod()) {// object key
			addToken(new TokenImpl(OP_MAP_PUSH, object));
			this.start++;// skip :
		} else if (isVar) {
			addToken(new TokenImpl(VALUE_VAR, object));
		} else {
			addToken(new TokenImpl(VALUE_CONSTANTS, object));
		}
	}

	private void addList() {
		addToken(new TokenImpl(BRACKET_BEGIN, null));
		addToken(new TokenImpl(VALUE_NEW_LIST, null));
		if (!skipSpace(']')) {
			addToken(new TokenImpl(OP_PARAM_JOIN, null));
		}
	}

	private void addMap() {
		addToken(new TokenImpl(BRACKET_BEGIN, null));
		addToken(new TokenImpl(VALUE_NEW_MAP, null));
	}
}

File: Lite/web/scripts/org/xidea/lite/expression-tokenizer.js
/*
 * List Template
 * License LGPL(您可以在任何地方免费使用,但请不要吝啬您对框架本身的改进)
 * http://www.xidea.org/project/lite/
 * @author jindw
 * @version $Id: template.js,v 1.4 2008/02/28 14:39:06 jindw Exp $
 */

//编译期间标记，compile time object
var BRACKET_BEGIN = 0xFFFE;//([{;
var BRACKET_END = 0xFFFF;//)]};

var STATUS_BEGIN = -100;
var STATUS_EXPRESSION = -101;
var STATUS_OPERATOR = -102;
var fns = {
	getResult :function() {
		return this.expression;//.slice(0).reverse();// reversed
	},

	parseEL :function() {
		this.skipSpace(0);
		while (this.start < this.end) {
			var c = this.value.charAt(this.start);
			if (c == '"' || c == '\'') {
				var text = this.findString();
				this.addKeyOrObject(text, false);
			} else if (c >= '0' && c <= '9') {
				var number = this.findNumber();
				this.addKeyOrObject(number, false);
			} else if (/[\w$_]/.test(c)) {
				var id = this.findId();
				switch(id){
				    case 'true':
				        this.addToken([VALUE_CONSTANTS,true]);
				        break;
				    case 'false':
				        this.addToken([VALUE_CONSTANTS,false]);
				        break;
				    case 'null':
				        this.addToken([VALUE_CONSTANTS,null]);
				        break;
				    default:
    				    this.skipSpace(0);
    					if (this.previousType == OP_GET_PROP) {
    						this.addToken([VALUE_CONSTANTS,
    								id]);
    					} else {
    						this.addKeyOrObject(id, true);
    					}
				}
			} else {
				var op = this.findOperator();
				// if (this.value.startsWith(op, this.start))
				this.parseOperator(op);
				if (op == null) {
					throw new Error("语法错误:" + this.value + "@"
							+ this.start);
				}
			}
			this.skipSpace(0);
		}
	},

	findOperator :function() {// optimize json ,:[{}]
		switch (this.value.charAt(this.start)) {
		case '!':// !,!=
		case '>':// >,>=
		case '<':// <,<=
			if (this.value.charAt(this.start + 1) == '=') {
				return this.value.substring(this.start, this.start += 2);
			}
		case ',':// optimize for json
		case ':':// 3op,map key
		case '[':// list
		case ']':
		case '{':// map
		case '}':
		case '(':// quote
		case ')':
		case '.':// prop
		case '?':// 3op
		case '+':// 5op
		case '-':
		case '*':
		case '/':
		case '%':
			return this.value.substring(this.start, this.start += 1);

		case '=':// ==
		case '&':// &&
		case '|':// ||
    		if(":debug"){
    			if (this.value.charAt(this.start) != this.value.charAt(this.start + 1)){
    			    $log.error("未知符号:" , this.value.substring(this.start),this.value)
    			}
    		}
			return this.value.substring(this.start, this.start += 2);
		}
		$log.error("未知符号",this.value,this.value.substring(this.start))
		return null;
	},

	/**
	 * 碰見:和,的時候，就需要檢查是否事map的間隔符號了
	 * 
	 * @return
	 */
	isMapMethod :function() {
		var i = this.tokens.length - 1;
		var depth = 0;
		for (; i >= 0; i--) {
			var token = this.tokens[i];
			var type = token[0];
			if (depth == 0) {
				if (type == OP_MAP_PUSH
						|| type == VALUE_NEW_MAP) {// (
					// <#newMap>
					// <#push>
					return true;
				} else if (type == OP_PARAM_JOIN) {// (
					// <#newList>
					// <#param_join>
					return false;
				}
			}
			if (type == BRACKET_BEGIN) {
				depth--;
			} else if (type == BRACKET_END) {
				depth++;
			}
		}
		return false;
	},

	parseOperator :function(op) {
		if (op.length == 1) {
			switch (op.charAt(0)) {
			case '(':
				if (this.status == STATUS_EXPRESSION) {
					this.addToken([OP_INVOKE_METHOD]);
					if (this.skipSpace(')')) {
						this.addToken([VALUE_CONSTANTS,
								[]]);
						this.start++;
					} else {
						this.addList();
					}

				} else {
					this.addToken([BRACKET_BEGIN]);
				}
				break;
			case '[':
				if (this.status == STATUS_EXPRESSION) {// getProperty
					this.addToken([OP_GET_PROP]);
					this.addToken([BRACKET_BEGIN]);
				}else {// list
					this.addList();
				}
				break;
			case '{':
				this.addMap();
				break;
			case '}':
			case ']':
			case ')':
				this.addToken([BRACKET_END]);
				break;
			case '+'://
				this.addToken([
						this.status == STATUS_EXPRESSION ? OP_ADD : OP_POS]);
				break;
			case '-':
				this.addToken([
						this.status == STATUS_EXPRESSION ? OP_SUB
								: OP_NEG]);
				break;
			case '?':// ?:
				this.addToken([OP_QUESTION]);
				// this.addToken(OperatorToken.getToken(SKIP_QUESTION));
				break;
			case ':':// :(object_setter is skiped)
				this.addToken([OP_QUESTION_SELECT]);
				break;
			case ',':// :(object_setter is skiped,',' should
				// be skip)
				if (!this.isMapMethod()) {
					this.addToken([OP_PARAM_JOIN]);

				}
				break;
			case '/':
				var next = this.value.charAt(this.start);
				if (next == '/') {
					var end1 = this.value.indexOf('\n', this.start);
					var end2 = this.value.indexOf('\r', this.start);
					var cend = Math.min(end1, end2);
					if (cend < 0) {
						cend = Math.max(end1, end2);
					}
					if (cend > 0) {
						this.start = cend;
					} else {
						this.start = this.end;
					}
					break;
				} else if (next == '*') {
					var cend = this.value.indexOf("*/", this.start);
					if (cend > 0) {
						this.start = cend + 2;
					} else {
						throw new Error("未結束注釋:" + this.value
								+ "@" + this.start);
					}
					break;
				}else if(this.status != STATUS_EXPRESSION){
					var end = findRegExp(this.value,this.start);
					if(end>0){
						this.addToken([VALUE_CONSTANTS,
							toValue(
								this.value.substring(this.start-1,end))]);
						this.start = end;
						break;
					}
				}
			default:
				this.addToken([findTokenType(op)]);
			}
		} else if (op == "||") { // ||
			this.addToken([OP_OR]);
			// this.addToken(LazyToken.LAZY_TOKEN_END);
		} else if (op == "&&") {// &&
			this.addToken([OP_AND]);
			// this.addToken(OperatorToken.getToken(SKIP_AND));
		} else {
			this.addToken([findTokenType(op)]);
		}
	},

	addToken :function(token) {
		switch (token[0]) {
		case BRACKET_BEGIN:
			this.status = STATUS_BEGIN;
			break;
		case VALUE_CONSTANTS:
		case VALUE_VAR:
		case BRACKET_END:
			this.status = STATUS_EXPRESSION;
			break;
		default:
			this.status = STATUS_OPERATOR;
			break;
		}
		// previousType2 = this.previousType;
		this.previousType = token[0];
		this.tokens.push(token);
	},

	addKeyOrObject :function(object, isVar) {
		if (this.skipSpace(':') && this.isMapMethod()) {// object key
			this.addToken([OP_MAP_PUSH, object]);
			this.start++;// skip :
		} else if (isVar) {
			this.addToken([VALUE_VAR, object]);
		} else {
			this.addToken([VALUE_CONSTANTS, object]);
		}
	},

	addList :function() {
		this.addToken([BRACKET_BEGIN]);
		this.addToken([VALUE_NEW_LIST]);
		if (!this.skipSpace(']')) {
			this.addToken([OP_PARAM_JOIN]);
		}
	},

	addMap :function() {
		this.addToken([BRACKET_BEGIN]);
		this.addToken([VALUE_NEW_MAP]);
	}
};
var pt = new JSONTokenizer('');
for(var n in fns){
    pt[n] = fns[n]
}
function toValue(v){
    return this.eval(v);
}
function findRegExp(text,start){
	var depth=0,c;
	while(c = text.charAt(start++)){
	    if(c=='['){
	    	depth = 1;
	    }else if(c==']'){
	    	depth = 0;
	    }else if (c == '\\') {
	        start++;
	    }else if(depth == 0 && c == '/'){
	    	while(c = text.charAt(start++)){
	    		switch(c){
	    			case 'g':
	    			case 'i':
	    			case 'm':
	    			break;
	    			default:
	    			return start-1;
	    		}
	    	}
	    	
	    }
	}
}
/**
 * 表达式解析器，将JS表达式文本解析成JSON中间代码
 */
function ExpressionTokenizer(value){
    this.value = value.replace(/^\s+|\s+$/g,'');
	this.start = 0;
	this.end = this.value.length;
    this.status = STATUS_BEGIN;
	this.previousType = STATUS_BEGIN;
	this.tokens = [];
	this.parseEL();
	this.expression = buildTree(trimToken(right(this.tokens)));
}
function buildTree(tokens){
	var stack = [];
    for(var i=0;i<tokens.length;i++){
        var item = tokens[i]
        var type = item[0];
        switch(type){
            case VALUE_CONSTANTS:
            case VALUE_VAR:
            case VALUE_NEW_LIST:
            case VALUE_NEW_MAP:
                stack.push(item);
                break;
            default://OP
                if(type & 1){//两个操作数
                    var arg2 = stack.pop();
                    var arg1 = stack.pop();
                    var el = [type,arg1,arg2]
                }else{//一个操作树
                	var arg1 = stack.pop();
                	var el = [type,arg1]
                }
                switch(type){
					case OP_GET_STATIC_PROP:
					case OP_INVOKE_METHOD_WITH_STATIC_PARAM:
					case OP_MAP_PUSH:
					el.push(item[1]);
                }
                stack.push(el)
        }
    }
    return stack[0];
}
ExpressionTokenizer.prototype = pt;



// 将中序表达式转换为右序表达式
function right(tokens) {
	var rightStack = [[]];
	var buffer = [];

	for (var i = 0;i<tokens.length;i++) {
		var item = tokens[i];
		if (item[0] > 0) {
			if (buffer.length == 0) {
				buffer.push(item);
			} else if (item[0] == BRACKET_BEGIN) {// ("(")
				buffer.push(item);
			} else if (item[0] == BRACKET_END) {// .equals(")"))
				while (true) {
					var operator = buffer.pop();
					if (operator[0] == BRACKET_BEGIN) {
						break;
					}
					addRightToken(rightStack, operator);
				}
			} else {
				while (buffer.length!=0
						&& rightEnd(item[0], buffer[buffer.length-1][0])) {
					var operator = buffer.pop();
					// if (operator[0] !=
					// BRACKET_BEGIN){
					addRightToken(rightStack, operator);
				}
				buffer.push(item);
			}
		} else {
			addRightToken(rightStack, item);
		}
	}
	while (buffer.length !=0) {
		var operator = buffer.pop();
		addRightToken(rightStack, operator);
	}
	return rightStack[rightStack.length-1];
}
function trimToken(tokens){
	for(var i=0;i<tokens.length;i++){
		var token = tokens[i];
		token.length = getTokenLength(token[0]);
	}
	return tokens;
}

function addRightToken(rightStack,
		token) {
	var list = rightStack[rightStack.length-1];
	if (token[0] == OP_GET_PROP) {
	    var last = list.length-1;
	    if(last>=0){
	        var previous = list[last];
	        if(previous[0] == VALUE_CONSTANTS){
	            list.length--;
	            token = [OP_STATIC_GET_PROP,previous[1]]; 
	        }
	    }
	}
	token.toString = toTokenString;
	list.push(token);
}

function getPriority(type) {
	switch (type) {
	case BRACKET_BEGIN:
	case BRACKET_END:
		return Math.MIN_VALUE;
	default:
		return (type & BIT_PRIORITY)<<4 | (type & BIT_PRIORITY_SUB)>>12;
	}
}
/**
 */
function rightEnd(currentType, priviousType) {
	var priviousPriority = getPriority(priviousType);
	var currentPriority = getPriority(currentType);
	//1?1:3 + 0?5:7 ==>1
	//1?0?5:7:3 ==>7
	//1?0?5:0?11:13:3 ==>13
	if(currentPriority <= priviousPriority){
		if(currentPriority == priviousPriority){
			if(currentType == OP_QUESTION_SELECT){
				return priviousType == OP_QUESTION;
			}else if(currentType == OP_QUESTION){
				return priviousType != OP_QUESTION_SELECT;
			}
		}
		return true;
	}else{
		return false;
	}
}